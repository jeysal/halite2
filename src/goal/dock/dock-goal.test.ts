import DockingStatus from 'halite/DockingStatus';
import GameMap from 'halite/GameMap';
import Planet from 'halite/Planet';
import Ship, { OwnShip } from 'halite/Ship';

import { emptyMap } from '../../test/util';

import DockGoal from './dock-goal';

describe('value', () => {
  test('is appropriate for a central planet', () => {
    expect(
      new DockGoal(new Planet(emptyMap, { x: 50, y: 50 })).value(emptyMap),
    ).toMatchSnapshot();
  });

  test('is appropriate for a remote planet', () => {
    expect(
      new DockGoal(new Planet(emptyMap, { x: 15, y: 15 })).value(emptyMap),
    ).toMatchSnapshot();
  });

  test('is high if nearby planets are already owned', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlanets([{ x: 10, y: 10, ownerId: 0 }]);
    gm.addPlanets([{ x: 90, y: 90, ownerId: 1 }]);
    expect(
      new DockGoal(new Planet(gm, { x: 15, y: 15 })).value(gm),
    ).toMatchSnapshot();
  });

  test('is low if no nearby planets are already owned', () => {
    const gm = new GameMap({ myPlayerId: 0, width: 100, height: 100 });
    gm.addPlanets([{ x: 10, y: 10, ownerId: 1 }]);
    gm.addPlanets([{ x: 90, y: 90, ownerId: 0 }]);
    expect(
      new DockGoal(new Planet(gm, { x: 15, y: 15 })).value(gm),
    ).toMatchSnapshot();
  });
});

describe('minShips', () => {
  test('is 1', () => {
    expect(new DockGoal(new Planet(emptyMap, {})).minShips()).toBe(1);
  });
});
describe('maxShips', () => {
  test('is equal to the number of docking spots', () => {
    expect(
      new DockGoal(new Planet(emptyMap, { dockingSpots: 3 })).maxShips(),
    ).toBe(3);
  });
});

describe('suitability', () => {
  test('is 1 for a ship already docked there', () => {
    expect(
      new DockGoal(new Planet(emptyMap, { id: 42 })).suitability(
        new Ship(emptyMap, 0, {
          dockingStatus: DockingStatus.DOCKED,
          dockedPlanetId: 42,
        }) as OwnShip,
        emptyMap,
      ),
    ).toBe(1);
  });

  test('is 0 for a ship docked elsewhere', () => {
    expect(
      new DockGoal(new Planet(emptyMap, { id: 42 })).suitability(
        new Ship(emptyMap, 0, {
          dockingStatus: DockingStatus.DOCKED,
          dockedPlanetId: 1337,
        }) as OwnShip,
        emptyMap,
      ),
    ).toBe(0);
  });

  test('is appropriate for a nearby ship', () => {
    expect(
      new DockGoal(new Planet(emptyMap, { id: 42, x: 15, y: 15 })).suitability(
        new Ship(emptyMap, 0, {
          dockingStatus: DockingStatus.UNDOCKED,
          x: 10,
          y: 10,
        }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });

  test('is appropriate for a distant ship', () => {
    expect(
      new DockGoal(new Planet(emptyMap, { id: 42, x: 15, y: 15 })).suitability(
        new Ship(emptyMap, 0, {
          dockingStatus: DockingStatus.UNDOCKED,
          x: 90,
          y: 90,
        }) as OwnShip,
        emptyMap,
      ),
    ).toMatchSnapshot();
  });
});

describe('strive', () => {
  test('docks at a dockable planet', () => {
    expect(
      new DockGoal(
        new Planet(emptyMap, {
          id: 42,
          x: 15,
          y: 15,
          radius: 4.5,
          dockingSpots: 3,
          dockedShipIds: [100],
        }),
      ).strive(new Ship(emptyMap, 0, {
        id: 105,
        dockingStatus: DockingStatus.UNDOCKED,
        x: 10,
        y: 10,
      }) as OwnShip),
    ).toMatchSnapshot();
  });

  test('navigates towards a non-dockable planet', () => {
    expect(
      new DockGoal(
        new Planet(emptyMap, {
          id: 42,
          x: 15,
          y: 15,
          radius: 4.5,
          dockingSpots: 3,
          dockedShipIds: [100],
        }),
      ).strive(new Ship(emptyMap, 0, {
        id: 105,
        dockingStatus: DockingStatus.UNDOCKED,
        x: 90,
        y: 90,
      }) as OwnShip),
    ).toMatchSnapshot();
  });
});
