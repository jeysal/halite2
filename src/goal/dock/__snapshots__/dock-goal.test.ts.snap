// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`strive docks at a dockable planet 1`] = `"d 105 42"`;

exports[`strive navigates towards a non-dockable planet 1`] = `"t 105 7 225"`;

exports[`suitability is appropriate for a distant ship 1`] = `0.25`;

exports[`suitability is appropriate for a nearby ship 1`] = `0.95`;

exports[`value is appropriate for a central planet 1`] = `0.6`;

exports[`value is appropriate for a remote planet 1`] = `0.53`;

exports[`value is high if nearby planets are already owned 1`] = `0.72`;

exports[`value is low if no nearby planets are already owned 1`] = `0.5800000000000001`;
